util/types/datum.go

```diff
@@ -791,18 +791,18 @@ func ProduceStrWithSpecifiedTp(s string, tp *FieldType, sc *variable.StatementCo
 				var truncateLen int
 				for i := range s {
 					if runeCount == flen {
-						// We don't break here because we need to iterate to the end to get runeCount.
 						truncateLen = i
+						break
 					}
 					runeCount++
 				}
-				err = ErrDataTooLong.Gen("Data Too Long, field len %d, data len %d", flen, runeCount)
+				err = ErrDataTooLong.Gen("Data Too Long, field len %d, data len %d", flen, len([]rune(s)))
```

Comment:
Why change here?

Reply:
runeCount == len([]rune(s)), not need to iterator to count it.

Reply:
But cast string to rune slice is more expensive.